### PR TITLE
[JENKINS-39300] View.getDisplayName() is ignored when rendering tabs

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-39300](https://issues.jenkins-ci.org/browse/JENKINS-39300)

Also:
- adds support for descriptor visibility filtering on views
- fixes l10n of AllView
- adds support for contextual create permission of specific view types
- some other small code fixes spotted while touching the files

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/View.java
Comment (stephenc) [inline core/src/main/java/hudson/model/View.java]: Note: this is the javadoc that applied to the method *before* my changes as the method would have blown up with a NPE if invoked outside of the `newView.jelly` screen of a `ViewGroup` as soon as `AllView.isInstantiable()` was called

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: This looks like it fixes JENKINS-38606.

File: core/src/main/resources/lib/layout/tab.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/tab.jelly]: Do we have an `@since` equivalent for Jelly tags?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: It is. But will it cause a fallout in upgrading Jenkins instances?

Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: Will only affect new instances. Old instances can get the fix by recreating the All view (add new view, change default, delete all view, add back all view, change default, tidy up)

Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: This looks like a reasonable default for new All views. No need to edit existing ones. Would just cause confusion and broken links.

File: core/src/main/resources/lib/layout/tab.jelly
Comment (stephenc) [inline core/src/main/resources/lib/layout/tab.jelly]: I am restoring behaviour that was broken in https://github.com/stephenc/jenkins/commit/2de93f861a8e6997d8f4b7b186b7953c835771cc#diff-1948650f90f2cefd48aff59fa5742aa0L58 (appears to be removed by omission as https://github.com/stephenc/jenkins/blob/2de93f861a8e6997d8f4b7b186b7953c835771cc/core/src/main/resources/lib/layout/tab.jelly#L29 was retained

**Comment by stephenc** [comment]
@daniel-beck @oleg-nenashev so I have added a fix for JENKINS-38606... and we need to be able to invoke that fix from plugins (e.g. folders plugin) so I cannot make that `@Restricted` before you ask!

**Comment by daniel-beck** [comment]
> have added a fix for JENKINS-38606

Not actually safe, as there are code paths that add the view name to the URL even if it's the default view.

I don't think we should fix JENKINS-38606 for existing installations. It modifies user data for a pretty weak reason.

File: core/src/main/resources/lib/layout/tab.jelly
Comment (stephenc) [inline core/src/main/resources/lib/layout/tab.jelly]: https://github.com/stapler/stapler/pull/84 created for you ;-)

**Comment by stephenc** [comment]
@daniel-beck OK, I have made this opt-in. I still think we should have the ability for an admin to force the fix applied everywhere as it could be quite difficult to fix all the `AllView`s scattered everywhere otherwise

**Comment by daniel-beck** [comment]
Is this really not trivial to take care of in a script console script with maybe five lines?

Do we really need a startup parameter (to be added in service configurations) for a one-time cleanup action here?

**Comment by stephenc** [comment]
@daniel-beck really? show me the five lines that find *every* ViewGroup (including the User properties for *all* users) and the folders etc and does the fix-up? For most admins it should not be an issue... but if it is an issue it could be very painful to fix by script or manually. Much better to provide a hook to fix the issue that is available to everyone IMHO.

**Comment by stephenc** [comment]
@daniel-beck can you respond please? Would like to get the change at least merged before we hit one of the last likely candidates for the next LTS

**Comment by daniel-beck** [comment]
I still have my doubt but I'd be happy to be outvoted. So far, nobody else has cared enough though  @jenkinsci/code-reviewers Ping!). FWIW I would like to see other opinions on solving this using a new parameter.

**Comment by rsandell** [comment]
:bee:

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]:
```java
@Restricted(NoExternalUse.class)
```

(and remove `@since`) unless this is something you really expect plugins to use.

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]:
```html
<p>
```

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: …though it _can_ be, and those URLs will break. Right?

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: `getBoolean(String)` suffices.

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: AFAICT `getViewName()` should be marked `@Nonnull`.

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: Should this not `save` the owner?

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java]: Perhaps just check `owner == null` and return `all()`?

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java]: `AccessDeniedException`, or better yet define a new `ACL.checkCreatePermission` overload for this.

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java]: ditto

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java]: Need to also check whether the `StaplerRequest` is null.

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: `request` could be null.

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: ditto

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: Is there a particular subclass you had in mind? Not in this PR; some other linked one?

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: Specify default implementation.

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: Specify default implementation.

File: core/src/main/resources/hudson/model/MyViewsProperty/newView.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/MyViewsProperty/newView.jelly]: Gratuitous diff hunk.

File: core/src/main/resources/jenkins/model/Jenkins/newView.jelly
Comment (jglick) [inline core/src/main/resources/jenkins/model/Jenkins/newView.jelly]: ditto

**Review by jglick — CHANGES_REQUESTED** [review]
Some minor inline comments, and [tests are failing](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/6417/testReport/); :+1: on concept.

File: core/src/main/java/hudson/model/AllView.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/AllView.java]: Yes, and there are links in Jenkins to the default view's `/view/whatever` URL.

File: core/src/main/java/hudson/model/AllView.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AllView.java]: I expect plugins that implement ViewGroup will need to call into this method

File: core/src/main/java/hudson/model/AllView.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AllView.java]: nope, responsibility of the caller

File: core/src/main/java/hudson/model/AllView.java
Comment (stephenc) [inline core/src/main/java/hudson/model/AllView.java]: Done

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (stephenc) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: Yes but it needs to use the `Referer` header hack to resolve the `View` so probably waiting on dedicated support in stapler for that

**Review by stephenc — COMMENTED** [review]
@jglick I think i have addressed all your comments (in some way or other)

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: Is there a proposed PR calling it? From `cloudbees-folder`? Then you should probably have a clearer name.

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: meaning you could use

```java
if (v.getViewName().equals(primaryView)) {
```

File: core/src/main/java/hudson/model/ViewDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewDescriptor.java]: So where is that subclass?

File: core/src/main/java/hudson/views/ListViewColumn.java
Comment (jglick) [inline core/src/main/java/hudson/views/ListViewColumn.java]: gratuitous reformatting

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: Making this disabled by default makes it less than useful.

IIUC the concern is about compatibility for direct links to, say, `http://jenkins/view/All/job/something/`. But that is easily solved: amend `ViewGroupMixIn.getView` (called from `Jenkins.getView(String)`, the Stapler URL binding) to fall back to searching by `View.getDisplayName` if no match is found via `getViewName`.

**Review by jglick — APPROVED** [review]
Would like to discuss having better URL compatibility, and would like to see the downstream PR that various API methods here imply.

**Comment by stephenc** [comment]
See https://github.com/jenkinsci/cloudbees-folder-plugin/pull/79 for example usage of `migrateLegacyPrimaryAllViewLocalizedName` method

**Comment by stephenc** [comment]
@rsandell @daniel-beck @jglick PR has been updated

File: core/src/main/java/hudson/model/ViewGroupMixIn.java
Comment (jglick) [inline core/src/main/java/hudson/model/ViewGroupMixIn.java]: Sounds right but hard for me at least to follow how this will work in progress; recommend having some code coverage using `@LocalData` to prove it works.

File: core/src/main/java/hudson/security/ACL.java
Comment (jglick) [inline core/src/main/java/hudson/security/ACL.java]: `FIXME`

**Review by jglick — APPROVED** [review]
Sounds good to the extent I understand it.

**Review by jglick — CHANGES_REQUESTED** [review]
Ah but lots of test failures.

**Comment by daniel-beck** [comment]
One test still fails:

	Expected: a string containing "ERROR: Jenkins does not allow to delete 'All' view"
	     but: was "
	ERROR: Jenkins does not allow to delete 'all' view
	"

File: core/src/main/java/hudson/model/ViewGroupMixIn.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/ViewGroupMixIn.java]: So changing the translations will break this compatibility measure for that language?

**Comment by stephenc** [comment]
I think we're good to merge now... i'll hold off until tomorrow and then merge just in case anyone has and issues with the test changes

File: core/src/main/java/hudson/model/AllView.java
Comment (amuniz) [inline core/src/main/java/hudson/model/AllView.java]: Yes, actually the new item link has switched from `/view/All/newJob` to `/view/all/newJob` (note the lowercase in the default view name).

File: core/src/main/java/hudson/model/AllView.java
Comment (jglick) [inline core/src/main/java/hudson/model/AllView.java]: #3088

**Pull request merged**
